### ISSUE TITLE
`NoReturn` on `__new__` and `__init__` disables accessing `@classmethod`s and `ClassVar`s directly from the class

### ISSUE THREAD

**Issue by bzoracler** [python/mypy#14044]
**Bug Report**

Annotating `__new__(...) -> typing.NoReturn` or `__init__(...) -> typing.NoReturn` causes false positives when accessing `@classmethod`s and `typing.ClassVar`-annotated attributes.

**To Reproduce**

```python
from typing import NoReturn, Any, ClassVar

class A:
    attr: ClassVar[str] = ""
    def __init__(cls, *args: Any, **kwargs: Any) -> NoReturn:
        raise RuntimeError
    @classmethod
    def f(cls) -> None:
        pass

class B:
    attr: ClassVar[str] = ""
    @classmethod
    def f(cls) -> None:
        pass
```

**Expected Behavior**

No issues on any of the following:
```python
A.f()
A.attr
B.f()
B.attr
```

**Actual Behavior**

```python
A.f()  # mypy: "Callable[[VarArg(Any), KwArg(Any)], NoReturn]" has no attribute "f"  [attr-defined]
A.attr  # mypy: "Callable[[VarArg(Any), KwArg(Any)], NoReturn]" has no attribute "attr"  [attr-defined]
B.f()
B.attr
```

**Your Environment**

- Mypy version used: 0.990 and master branch of https://mypy-play.net/
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10

**Comment by dmcooke** [comment]
Just hit this bug. It's not just `@classmethod` or `ClassVar`: it'll complain about class decorators, ordinary methods, and staticmethod's.  Using `reveal_type`, it appears that mypy thinks that `A` is a function (a `Callable[..., NoReturn]`), 
and has lost all information about `A` being a class.

**Comment by fofoni** [comment]
Just hit this too.

> Just hit this bug. It's not just `@classmethod` or `ClassVar`: it'll complain about class decorators, ordinary methods, and staticmethod's. Using `reveal_type`, it appears that mypy thinks that `A` is a function (a `Callable[..., NoReturn]`), and has lost all information about `A` being a class.

Actually, this is apparently not related to this bug:

```python
class B:
    pass
reveal_type(B)  # Revealed type is "def () -> __main__.B"
```

Anyway, I can confirm that this bug still exists on mypy 1.6.1 and currently on mypy master branch, and occurs both on Python 3.11 and Python 3.12. I can also confirm that pyright doesn't have this bug (i.e. it allows `__init__(self) -> NoReturn` and then accessing classmethods).

**Comment by lkct** [comment]
Also hit this. I confirm this bug still persists in 1.7.1 and the master branch, on python versions 3.8 through 3.12.

Is anyone looking into this issue? I find it useful sometimes to define a class that does not instantiate while keeping ClassVars and classmethods available.

**Comment by henzef** [comment]
This is most likely caused by https://github.com/python/mypy/pull/11216

**Comment by BvB93** [comment]
I'm afraid not. My familiarity with mypy's internals is limited, even more so here as this concerns 2.5 year old PRs.
